client/src/unifycr-sysio.c

```diff
@@ -1523,13 +1546,24 @@ ssize_t UNIFYCR_WRAP(pread)(int fd, void *buf, size_t count, off_t offset)
     /* check whether we should intercept this file descriptor */
     if (unifycr_intercept_fd(&fd)) {
         /* get pointer to file descriptor structure */
+        unifycr_filemeta_t *meta = unifycr_get_meta_from_fid(fd);
         unifycr_fd_t *filedesc = unifycr_get_filedesc_from_fd(fd);
-        if (filedesc == NULL) {
+
+        if (meta == NULL || filedesc == NULL) {
             /* ERROR: invalid file descriptor */
             errno = EBADF;
             return (ssize_t) (-1);
         }
 
+        if (offset == meta->size)
+            return 0;
+
+        /* we should set errno conforming to lseek(2) */
+        if (offset > meta->size) {
```

Comment:
Haven't tested pread in this case, but it looks like lseek allows for seeking past the end of a file:

"The lseek() function allows the file offset to be set beyond the end of the file (but this does not change the size of the file).  If data is later written at this point, subsequent reads of the data in the gap (a "hole") return null bytes ('\0') until data is actually written into the gap."

Perhaps just returning EOF is enough in this case?

Reply:
I will check with the `client/src/unifycr-dirops.c`. I remember I had a build error for that, and probably I have missed to put the fix together.

For the seeking the beyond of the file, `pread(2)` seems to allow reading beyond the length by simply returning 0. I have tested with the following program:

```
[root@ninja82 pread]$ cat test-pread.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <assert.h>
#include <errno.h>
#include <string.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

static const char *testfile = "passwd";

int main(int argc, char **argv)
{
    int ret = 0;
    int fd = 0;
    off_t offset = 0;
    struct stat sb = { 0, };
    char buf[512] = { 0, };

    fd = open(testfile, O_RDONLY);
    assert(fd > 0);

    ret = fstat(fd, &sb);
    assert(ret == 0);

    printf("size = %lu\n", sb.st_size);
    offset = sb.st_size * 2;

    ret = lseek(fd, SEEK_SET, offset);
    printf("lseek(offset=%lu): ret = %d, errno = %d, %s\n", offset, ret, errno, strerror(errno));

    errno = 0;
    ret = pread(fd, buf, 512, offset);
    printf("pread(offset=%lu): ret = %d, errno = %d, %s\n", offset, ret, errno, strerror(errno));

    close(fd);

    return ret;
}
[root@ninja82 pread]$ ./test-pread 
size = 1554
lseek(offset=3108): ret = -1, errno = 22, Invalid argument
pread(offset=3108): ret = 0, errno = 0, Success
```

I will fix these things and submit again. Thanks!

Reply:
Thanks for the test, @sandrain.  That helps.  It surprised me a bit that lseek() was returning an error here.  I think that works as well, but we've got the argument order mixed in the test code above.  If I change:
```
ret = lseek(fd, SEEK_SET, offset);
```
to:
```
ret = lseek(fd, offset, SEEK_SET);
```
then it outputs:
```
size = 122398
lseek(offset=244796): ret = 244796, errno = 0, Success
pread(offset=244796): ret = 0, errno = 0, Success
```

Reply:
sorry, my bad :-)